### ISSUE TITLE
Interpreter crash on literal dictionary inside literal list.

### ISSUE THREAD

**Issue by lnihlen** [supercollider/supercollider#6162]
## Environment

* SuperCollider version: 3.13.0
* Operating system: macOS Sonoma 14.2
* Other details (Qt version, audio driver, etc.):
Apple Silicon

## Steps to reproduce

```supercollider
#[()]
```

## Expected vs. actual behavior

Interpreter crash on evaluation.

[sclang-2023-12-15-185539.ips.zip](https://github.com/supercollider/supercollider/files/13691091/sclang-2023-12-15-185539.ips.zip)

I can investigate but it will likely be a few days before I generate a patch.

**Comment by muellmusik** [comment]
Not that it should crash, but I don't think literal dictionaries are supported.

**Comment by lnihlen** [comment]
You're right, technically they aren't dictionaries, they are `Event`s:

```
(
var e = ();
e.class.postln;   // prints Event 

e = (\foo: 1, \bar: 2);
e.class.postln;   // prints Event
)
```

I just call them dictionaries because *almost all* of the sclang internals does.

**Comment by muellmusik** [comment]
Yes. Though I meant Dictionary in the OOP sense, as Event is a Dictionary subclass.

```
Event.superclasses
-> [ class Environment, class IdentityDictionary, class Dictionary, class Set, class Collection, class Object ]

#[Dictionary[]]
ERROR: Only Array is supported as literal type.
Compiling as an Array.
```
But IIRC () = an event is a bit of syntactic sugar and is probably not treated as literal?

Still shouldn't crash of course, but I thought that might be a hint.

**Comment by lnihlen** [comment]
The problem with the `#[Dictionary[]]` statement isn't the fact that it's in a list literal, it's the fact that you can't declare a non-Array class as using the `ClassName[]` syntax:

```
#Dictionary[]

ERROR: Only Array is supported as literal type.
Compiling as an Array.
-> [  ]
```

I understand the importance of the `#` here in determining if something is truly a *literal* or no, but my reading of the sclang grammar is that while there is no support for bare literal Dictionaries, support was at least intended for literal dictionaries inside of literal arrays. Couple of points from the (edited here for clarity) sclang grammar:

```bison
expr1	: pushliteral
// <snip>
	|  '[' arrayelems ']'
		{ $$ = (intptr_t)newPyrDynListNode(0, (PyrParseNode*)$2); }
	|	'(' dictslotlist ')'
		{ $$ = (intptr_t)newPyrDynDictNode((PyrParseNode*)$2); }
        ;


pushliteral : integer
               { $$ = (intptr_t)newPyrPushLitNode((PyrSlotNode*)$1, NULL); }
// <snip>
	    | listlit	{ $$ = (intptr_t)newPyrPushLitNode(NULL, (PyrParseNode*)$1); }

listlit : '#' '[' literallistc ']'
              { $$ = (intptr_t)newPyrLitListNode(0, (PyrParseNode*)$3); }
        | '#' classname  '[' literallistc ']'
              { $$ = (intptr_t)newPyrLitListNode((PyrParseNode*)$2, (PyrParseNode*)$4); }
        ;
```

So here we see that `#[]` matches against `expr1` -> `pushliteral` -> `listlit` and results in the `PyrLitListNode`, ultimately resulting in a recompiled `Array` with literal semantics, whereas `[]` and `()` match against their dynamic forms in `expr1` resulting in `PyrDynListNode` and `PyrDynDictNode` respectively, and the compiled code will create new elements and then add the contents inside, which match against general `exprSeq` and so can be determined at runtime.

But, if we dig into `literallistc` we see:

```bison
literallistc	: { $$ = 0; }
		| literallist1 optcomma
		;

literallist1	: listliteral
		| literallist1 ',' listliteral
	           { $$ = (intptr_t)linkNextNode((PyrParseNode*)$1, (PyrParseNode*)$3); }
		;


listliteral	: integer
                  { $$ = (intptr_t)newPyrLiteralNode((PyrSlotNode*)$1, NULL);
// <snip>
 		| listlit2
                   { $$ = (intptr_t)newPyrLiteralNode(NULL, (PyrParseNode*)$1); }
		| dictlit2 
                   { $$ = (intptr_t)newPyrLiteralNode(NULL, (PyrParseNode*)$1); }
		;
``` 

So `literallistc` is accepting zero or more `listliteral` elements, separated by commas, with an optional comma at the end. And two of the elements that match `listliteral` are `listlit2` and `dictlit2`:

```bison
listlit2 : '[' literallistc ']'
	    { $$ = (intptr_t)newPyrLitListNode(0, (PyrParseNode*)$2); }
	 | classname  '[' literallistc ']'
	    { $$ = (intptr_t)newPyrLitListNode((PyrParseNode*)$1, (PyrParseNode*)$3); }
	 ;


dictlit2 : '(' litdictslotlist ')'
	    { $$ = (intptr_t)newPyrLitDictNode((PyrParseNode*)$2); }
```

This is the *only* occurrence of `newPyrLitDictNode` in the *entire* sclang grammar. I'll spare you the details of `litdictslotlist`, suffice to say it does the usual song-and-dance about recursive parsing of comma-separated optional elements, ultimately matching pairs of  literals separated by a colon, or a keyword (sugar for a symbol) and a literal.

So, at least the grammar was built to support literal dictionaries inside of literal arrays. Seems like a niche use case but the code is there.

**Comment by muellmusik** [comment]
Interesting! The error seems to go back to the initial revision, FWIW.
